File: src/lib/player-auto-cross.js
/**
 * Pure helper for the master→player auto-cross path. Reads the master's
 * full `called[]` history (not a single-slot bus) so player regen,
 * "Xoá đánh dấu", and reload can replay missed draws by passing
 * `lastHandledIndex: 0`.
 *
 * Cursor advances strictly even when no cell flips (mode mismatch,
 * manual untick, off-board) so the same draw never re-fires.
 *
 * @module lib/player-auto-cross
 */

import { findUncrossedCell } from "$lib/game-logic.js";

/**
 * @param {object} args
 * @param {number[][] | null} args.grid
 * @param {boolean[][]} args.crossed
 * @param {number[]} args.called               - master's full history
 * @param {number} args.lastHandledIndex       - index already consumed
 * @param {Set<number>} args.manualUnticks     - numbers user explicitly unticked
 * @param {"player" | "master" | "both"} args.mode
 * @returns {{ crossed: boolean[][], lastHandledIndex: number, changed: boolean }}
 */
export function applyMasterCalls({
  grid,
  crossed,
  called,
  lastHandledIndex,
  manualUnticks,
  mode,
}) {
  if (lastHandledIndex >= called.length) {
    return { crossed, lastHandledIndex, changed: false };
  }
  // In non-both modes the player isn't auto-following the master, but we
  // still advance the cursor — otherwise toggling player→both would dump
  // the entire back-history at once. Only future draws should auto-cross.
  if (mode !== "both" || !grid || crossed.length === 0) {
    return { crossed, lastHandledIndex: called.length, changed: false };
  }
  let next = crossed;
  let changed = false;
  for (let i = lastHandledIndex; i < called.length; i++) {
    const num = called[i];
    if (manualUnticks.has(num)) continue;
    const target = findUncrossedCell(grid, next, num);
    if (!target) continue;
    next = next.map((row, ri) =>
      ri === target.row
        ? row.map((v, ci) => (ci === target.col ? true : v))
        : row,
    );
    changed = true;
  }
  return { crossed: next, lastHandledIndex: called.length, changed };
}

File: android/app/src/main/java/com/miti99/loto/game/PlayerAutoCross.kt
package com.miti99.loto.game

import com.miti99.loto.settings.AppMode

/**
 * Pure helper for the master→player auto-cross path, ported one-to-one from
 * `web/src/lib/player-auto-cross.js`. Reads the master's full call history
 * (not a single-slot bus) so player regen, "Xoá đánh dấu", and restore can
 * replay missed draws by passing `lastHandledIndex = 0`.
 *
 * The cursor advances strictly even when no cell flips (mode mismatch,
 * manual untick, off-board) so the same draw never re-fires.
 */
object PlayerAutoCross {

    data class Result(
        val crossed: List<List<Boolean>>,
        val lastHandledIndex: Int,
        val changed: Boolean,
    )

    fun applyMasterCalls(
        grid: List<List<Int>>?,
        crossed: List<List<Boolean>>,
        called: List<Int>,
        lastHandledIndex: Int,
        manualUnticks: Set<Int>,
        mode: AppMode,
    ): Result {
        // Hardening beyond the web: a shrunken history means the round was
        // reset, so a stale cursor restarts from 0 instead of silently
        // disabling auto-cross for the rest of the new round.
        val cursor = if (lastHandledIndex > called.size) 0 else lastHandledIndex
        if (cursor >= called.size) {
            return Result(crossed, cursor, changed = false)
        }
        // In non-both modes the player isn't auto-following the master, but
        // the cursor still advances — otherwise toggling player→both would
        // dump the entire back-history at once. Only future draws auto-cross.
        if (mode != AppMode.BOTH || grid == null || crossed.isEmpty()) {
            return Result(crossed, called.size, changed = false)
        }
        var next = crossed
        var changed = false
        for (i in cursor until called.size) {
            val num = called[i]
            if (manualUnticks.contains(num)) continue
            val target = PlayerCard.findUncrossedCell(grid, next, num) ?: continue
            next = next.mapIndexed { ri, row ->
                if (ri == target.row) {
                    row.mapIndexed { ci, v -> if (ci == target.col) true else v }
                } else {
                    row
                }
            }
            changed = true
        }
        return Result(next, called.size, changed)
    }
}
